datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+    eval_results,
+)
+from datalad.interface.results import annexjson2result
+from datalad.support.annexrepo import (
+    AnnexJsonProtocol,
+    AnnexRepo,
+)
+from datalad.support.gitrepo import GitRepo
+from datalad.support.param import Parameter
+from datalad.support.constraints import (
+    EnsureStr,
+    EnsureNone,
+    EnsureChoice,
+)
+from datalad.support.exceptions import CommandError
+from datalad.utils import (
+    Path,
+    ensure_list,
+)
+
+from datalad.distribution.dataset import (
+    Dataset,
+    EnsureDataset,
+    datasetmethod,
+    require_dataset,
+    resolve_path,
+)
+from datalad.core.local.diff import diff_dataset
+
+
+lgr = logging.getLogger('datalad.core.distributed.push')
+
+
+@build_doc
+class Push(Interface):
+    """Push a dataset to a known :term:`sibling`.
+
+    This makes the last saved state of a dataset available to a sibling
+    or special remote data store of a dataset. Any target sibling must already
+    exist and be known to the dataset.
+
+    Optionally, it is possible to limit a push to change sets relative
+    to a particular point in the version history of a dataset (e.g. a release
+    tag). By default, the state of the local dataset is evaluated against the
+    last known state of the target sibling. An actual push is only attempted
+    if there was a change compared to the reference state, in order to speed up
+    processing of large collections of datasets. Evaluation with respect to
+    a particular "historic" state is only supported in conjunction with a
+    specified reference dataset. Change sets are also evaluated recursively, i.e.
+    only those subdatasets are pushed where a change was recorded that is
+    reflected in to current state of the top-level reference dataset.
+    See "since" option for more information.
+
+    Only a push of saved changes is supported.
+
+    .. note::
+      Power-user info: This command uses :command:`git push` and :command:`git annex copy`
+      to push a dataset. Publication targets are either configured remote
+      Git repositories or git-annex special remotes (if they support data
+      upload).
+    """
+
+    # TODO add examples
+
+    _params_ = dict(
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc=""""specify the dataset to push""",
+            constraints=EnsureDataset() | EnsureNone()),
+        to=Parameter(
+            args=("--to",),
+            metavar='SIBLING',
+            doc="""name of the target sibling. If no name is given, an attempt is
+            made to identify the target based on the dataset's configuration
+            (i.e. a configured tracking branch, or a single sibling that is
+            configured for push)""",
+            constraints=EnsureStr() | EnsureNone()),
+        since=Parameter(
+            args=("--since",),
+            constraints=EnsureStr() | EnsureNone(),
+            doc="""specifies commit (treeish, tag, etc.) from which to look for
```

Comment:
`since` is passed to `get_hexsha()`, and that accepts any commit-ish but will fail on a plain tree object.

Reply:
done